#167392449 Implement Search Functionality

kechyy opened the pull request:
### What does this PR do?
- Implement search functionality

#### Description of Task to be completed?
- update user profile model with firstname and lastname
- create search functionality API controller
- create tag model
- create articletag model
- write unit tests to test the search functionality

#### How should this be manually tested?
- clone the repo
- run npm install
- start the server with npm run start:dev
- test on postman http://localhost:3000/api/v1/search?search='searchValue'

#### Any background context you want to provide?
- N/A

#### What are the relevant pivotal tracker stories?
- [#167392449](https://www.pivotaltracker.com/story/show/167392449)

#### Screenshots (if appropriate)

#### Questions:
- N/A

timi-codes reviewed (changes requested):
Good Job ladies 😂. I'm aware this PR is close to perfect 😇but I just have to request this changes because you forgot to document your API's.😎

Onyimatics commented:
> Good Job ladies 😂. I'm aware this PR is close to perfect 😇but I just have to request this changes because you forgot to document your API's.😎

 😂Thank you @timi-codes . We've added the documentation. Kindly re-review and do the needful. 😂 😂

codexempire reviewed (approved):
Nice one 👍 great algorithm

segunolalive commented on src/test/search.spec.js:
```diff
@@ -0,0 +1,151 @@
+  it('should search for articles by tag', async () => {
+    const firstUser = await createUser();
+    const firstArticle = await createArticle(
+      await generateArticle({ authorId: firstUser.id })
```
```

const article = await generateArticle({ authorId: firstUser.id });
article.title = 'bush gehs';
const secondarticle = await generateArticle({ authorId: firstUser.id });
secondarticle.title = 'zzzz';

const firstArticle = await createArticle(article);

// send 'bush gehs' as your search query;

now you can assert that the title_scrore === 1


```

segunolalive replied:
create a mock that you know will not match the query in any way and assert that it didn't match. That's how you know your search doesn't just match everything in the db

Onyimatics replied:
All right, we will do that.

kechyy replied:
> create a mock that you know will not match the query in any way and assert that it didn't match. That's how you know your search doesn't just match everything in the db

Okay, we'll implement the changes as requested. 

Thank you

segunolalive reviewed (changes requested):
Good work. Attend to the comments.

Consider adding pagination to this.

Onyimatics commented:
> Good work. Attend to the comments.
> 
> Consider adding pagination to this.

Okay, we will attend to them. Thank you.

tolumide-ng reviewed (approved):
Great job...welldone guys 👍

Commit:
167392449-feat(search functionality): implement search functionality

- update user profile model with firstname and lastname
- create search functionality API controller
- create tag model
- create articletag model

[Delivers #167392449]
